assets/shaders/custom_vertex_attribute.wgsl

```diff
@@ -23,7 +23,7 @@ struct VertexOutput {
 fn vertex(vertex: Vertex) -> VertexOutput {
     var out: VertexOutput;
     out.clip_position = mesh_position_local_to_clip(
-        mesh[bevy_render::instance_index::get_instance_index(vertex.instance_index)].model,
+        affine_to_square(mesh[bevy_render::instance_index::get_instance_index(vertex.instance_index)].model),
```

Comment:
> i think perhaps we should embrace the affinity and make the 3x4 the arg to the `mesh_xxx` and `skin_normal` functions, and the return of the `skin_model` function. maybe not very important since its only used per vertex but it would be cleaner.

Then each of those would have to do the conversions or be less-readable code, and if anyone wanted to use the transforms for something else then they would have to figure out how.

I am rather feeling like we may want to add a `get_model_matrix(instance_index) -> mat4x4<f32>` function.

> then this would be clearer/easier to understand if the mesh.model (and the `mesh_xxx` arg) was a mat4x3 instead of a transpose ... is there a good reason to use 3x4 instead? the packing logic in mesh.rs is already quite complex so it probably wouldn't be worse. i guess `MeshUniform` would contain a [f32;12] instead of a [Vec4;3].

4x3 takes the same space in uniform/storage buffers as 4x4 because it's practically 3 vec3s, and vec3s have an alignment of 16 so each vec3 starts at a 16-byte alignment. The shader would then error or misinterpret the data in the binding.

> the downside would be if a caller has a 4x4 and wants to use the 4x3 functions they would have to use `mesh_xxx(mat4x3(my4x4[0].xyz, my4x4[1].xyz, my4x4[2].xyz, my4x4[3].xyz))` - i guess we could add a utility conversion function for that if there isn't a better builtin way (i don't think there is).

Projection matrices are necessarily 4x4. so there we have to have a 4x4 matrix. One thing to note is that I have been told that the shader compiler (the GPU vendor one I guess?) will optimise away the multiplications by 0.0 and 1.0, so adding them back to make it a 4x4 does not hurt performance, it only makes code more readable and the matrix more usable.

Reply:
I have added a `get_model_matrix(instance_index: u32) -> mat4x4<f32>` helper function and similar for previous model matrix.

Reply:
everything makes sense given the size of mat4x3 is 64 bytes, i didn't realise that. then there's no purpose to 4x3 as input to the functions, and we would lose the size benefit if it was the unfiorm type. 👍